File: Python/0204-e-count-primes.py
"""
DP
Time: O(n)
Space: O(n)
"""
import math
class Solution:
    def countPrimes(self, n: int) -> int:
        if n <= 2:
            return 0
        
        notPrime = [False] * n
        bound = math.sqrt(n)
        count = 0
        for i in range(2,n):
            if not notPrime[i]:
                count += 1
            if i <= bound:
                for j in range(i**2, n, i):
                    if not notPrime[j]:
                        notPrime[j] = True
        return count

File: leetcode/C++/0204-e-count-primes.cpp
class Solution {
public:
    int countPrimes(int n) {
        if (n == 0) return 0;
        vector<bool> prime(n, true);
        prime[0] = false, prime[1] = false;
        for (int i = 2; i * i < n; i++) {
            if (prime[i]) {
                for (int j = i * i; j < n; j += i) {
                    prime[j] = false;
                }
            }
        }
        int cnt = 0;
        for (bool isPrime : prime)
            if (isPrime)
                cnt++;
        return cnt;
    }
};